### ISSUE TITLE
How to encode *NoMatch* after consumed some input in a parser

### ISSUE THREAD

**Issue by complyue** [complyue/dcp#2]
I just now kinda get to know some difference between `empty` vs `fail` in a parser, and hope that `empty` can be used to encode the semantic for whitespaces (including comments and doc comments just before eof) get consumed without a match.

Tried sth like this:

https://github.com/complyue/dcp/blob/5be688396b7e2bda00ea80fd99d2a7b3ec5c788d/src/Parser.hs#L138-L143

But it doesn't work as expected:

```console
$ cabal run dcp < samples/full.txt
Up to date
dcp: 73:1:
   |
73 | <empty line>
   | ^
expecting "{#", "{##", or artifact body

CallStack (from HasCallStack):
  error, called at src/Parser.hs:151:14 in main:Parser
```

So how to achieve that?

**Comment by complyue** [comment]
close this in favor of https://github.com/mrkkrp/megaparsec/issues/429

**Comment by complyue** [comment]
Solved by 39039e62db5f63e61814383ffc2c5f4fba240806 with help from Olaf

**Comment by complyue** [comment]
With help from Olaf, I think I just cracked my confusion now, that I failed
to distinguish between the 2 monoid structures of concatenation and
choice, and have wrongly assumed somehow Alternative can do
concatenation at the same time as choice making.

Now I understand it clearly that parsers can be concatenated only if
their result type is concatenatable, so if the combinator lib is to
address my demand, it can define `maynTill'` with the parser type
fixed like:

```hs
manyTill' :: Monoid a => Parser a -> Parser end -> Parser a
manyTill' p end = go where go = (mempty <$ end) <|> liftA2 mappend p go
```

Then I can tune my result types, and implement my parsers like:

```hs
type ArtDecl = (Maybe DocComment, Text)
data ModuleDecl m = (Applicative m, Monoid (m ArtDecl)) =>
  ModuleDecl (Maybe DocComment) (m ArtDecl)

moduleDecl :: (Applicative m, Monoid (m ArtDecl)) => Parser (ModuleDecl m)
moduleDecl = lexeme $ do
  sc
  moduCmt <- optional docComment
  arts    <- manyTill' (scWithSemiColons >> artifactDecl) eof
  return $ ModuleDecl moduCmt arts

artifactDecl :: (Applicative m, Monoid (m ArtDecl)) => Parser (m ArtDecl)
artifactDecl = lexeme $ do
  artCmt <- optional immediateDocComment
  choice
    [ eof >> return mempty
    , do
      artBody <- takeWhileP (Just "artifact body")
                            (not . flip elem (";{" :: [Char]))
      return $ pure (artCmt, artBody)
    ]
```

https://github.com/complyue/dcp/blob/39039e62db5f63e61814383ffc2c5f4fba240806/src/Parser.hs#L128-L152

This approach clears my doubts, though doesn't seem more ergonomic compared to approaches like

  https://github.com/complyue/dcp/commit/2600179ab480b2778589495c1ab0bb092f3ae7df

or 

  https://github.com/complyue/dcp/commit/f0c12d51e30cfc49e9e580e5e766c4e8f7d4b54e

Well, then I know my options now.
